lib/autoproj/daemon/github_watcher.rb

```diff
@@ -1,189 +1,224 @@
+
+            # @param [Github::PushEvent] push_event
+            # @return [PackageRepository, nil]
+            def package_by_push_event(push_event)
+                packages.find do |pkg|
+                    pkg.owner == push_event.owner &&
+                        pkg.name == push_event.name &&
+                        pkg.branch == push_event.branch
+                end
+            end
+
+            # @param [Github::PushEvent] push_event
+            # @return [void]
+            def dispatch_push_event(push_event)
+                to_mainline = to_mainline?(
+                    push_event.owner, push_event.name, push_event.branch
+                )
+                if to_mainline
+                    package = package_by_push_event(push_event)
+                    return if package.head_sha == push_event.head_sha
+                else
+                    cached_pull_request = cache.pull_requests.find do |pr|
+                        pr.head_owner == push_event.owner &&
+                            pr.head_name == push_event.name &&
+                            pr.head_branch == push_event.branch
                     end
-                end.size
+                    return unless cached_pull_request
+
+                    pull_request = client.pull_requests(
+                        cached_pull_request.base_owner,
+                        cached_pull_request.base_name,
+                    ).find { |pr| pr.number == cached_pull_request.number }
+                    return unless pull_request
+                end
+                call_push_hooks(push_event, mainline: to_mainline,
+                                            pull_request: pull_request)
             end
-            # rubocop: enable Metrics/AbcSize
 
-            # Whether an event is valid for a given repository
-            # A valid event is either a Pull Request or Push that
-            # are newer than the last handled event
-            #
-            # @return [Boolean]
-            def valid_event?(repo, event)
-                event_types = %w[PullRequestEvent PushEvent]
-                return false unless event_types.any? { |e| e == event['type'] }
-                return false if Time.parse(event['created_at'].to_s) < repo.timestamp
+            # @param [Array] events An array of events
+            # @return [void]
+            def handle_owner_events(events)
+                events = filter_events(events)
+                events = partition_events_by_repo_name(events)
+                events.each_value do |repo_events|
+                    push_events, pull_request_events = partition_events_by_type(
+                        repo_events
+                    )
 
-                true
+                    last_push_event = push_events.max_by(&:created_at)
+                    last_pull_request_event =
+                        pull_request_events.max_by(&:created_at)
+
+                    dispatch_push_event(last_push_event) if last_push_event
```

Comment:
You are filtering on date, but then you are (in `dispatch_push_event`) having other filters and two different codepaths. Doesn't that mean that `last_push_event` could be ignored while some other events would not be ?

Reply:
Damn... The rationale was that I don't care if a PR was opened and closed several times, I'm only interested in the last event (to find out whether the PR is currently open or closed). Same goes for push (I'm only interested in the last one). The problem is that I am doing this for the entire repo when I should be doing per PR (or per branch in the case of pushes). I have to partition this further to get it right :(

Do you have a better idea to implement this?

Reply:
I think you need to adopt more of a data processing pipeline structure. That is, having the processing done as passing data structure(s) from functions to functions until you have the data you need to call the hooks.

One method that basically does

~~~
[get events] -> [filter/process/transform all events] -> ["collapse" redundant events] -> call hooks
~~~

where the get/filter/transform/collapse/call are processing methods implemented here and there. Most of the classes would become "passive", that is not call anything that would end up having the system actually act.

That is, in the end, making sure that the "call hooks" is done at the first level of the stack, not the last. Basically the style I mentioned in one of the first pull requests for autoproj-daemon if you remember.

Reply:
@doudou Is this what you mean: https://github.com/rock-core/autoproj-daemon/pull/12/commits/4b1ce9205b79f412b9c11a0f66e7bd95a73a92ab ?